flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -396,21 +364,55 @@ synchronized void resetReceivedBarriers(long checkpointId) {
 			return allBarriersReceivedFuture;
 		}
 
-		synchronized void onChannelClosed() {
+		synchronized void onChannelClosed() throws IOException {
 			numOpenChannels--;
+
+			if (resetPendingCheckpoint()) {
+				handler.notifyAbort(
+					currentReceivedCheckpointId,
+					new CheckpointException(CheckpointFailureReason.CHECKPOINT_DECLINED_INPUT_END_OF_STREAM));
+			}
+		}
+
+		synchronized boolean setCancelledCheckpointId(long cancelledId) {
+			if (currentReceivedCheckpointId > cancelledId || (currentReceivedCheckpointId == cancelledId && numBarriersReceived == 0)) {
```

Comment:
`&& numBarriersReceived == 0` -> `!isCheckpointPending()`?

Reply:
My intention to not rely on `isCheckpointPending()` for new methods, also tried to get ride of it in previous core codes and make `isCheckpointPending()` only for test purpose.

Reply:
I think it also makes sense to use `isCheckpointPending()` considering this comment reason https://github.com/apache/flink/pull/12460#discussion_r437217966.

Maybe I want to refactor it next time when touching this code, since I do not want to re-execute the azure to block something. :)